### PR TITLE
Replace compile_error/4 with form_error/4 in elixir_erl_compiler

### PR DESCRIPTION
**Pull request by whatyouhide**

### CODE REVIEW

File: lib/elixir/src/elixir_erl_compiler.erl [hunk 1]
```diff
@@ -75,11 +75,12 @@ handle_file_warning(_, File, {Line, Module, Desc}) ->
 %% Handle warnings
 
 handle_file_error(File, {beam_validator, Rest}) ->
-  Message = beam_validator:format_error(Rest),
-  elixir_errors:compile_error([{line, 0}], File, elixir_utils:characters_to_binary(Message));
+  elixir_errors:form_error([{line, 0}], File, beam_validator, Rest);
 handle_file_error(File, {Line, Module, Desc}) ->
-  Message = format_error(Module, Desc),
-  elixir_errors:compile_error([{line, Line}], File, elixir_utils:characters_to_binary(Message)).
+  elixir_errors:form_error([{line, Line}], File, ?MODULE, {Module, Desc}).
```
Comment (josevalim) [inline lib/elixir/src/elixir_erl_compiler.erl]: Hrm, not a fan of the indirection on this case. Thoughts?

Comment (whatyouhide) [inline lib/elixir/src/elixir_erl_compiler.erl]: Not a fan either, but I want to get rid of `compile_error` completely now :P Suggestions?

Comment (josevalim) [inline lib/elixir/src/elixir_erl_compiler.erl]: Yes. Don't get rid of `compile_error` completely. :P

Comment (whatyouhide) [inline lib/elixir/src/elixir_erl_compiler.erl]: Oh but that sucks 😄 If I'm correct this is the only place where we still use `compile_error`. I would push for finding a better way to abstract this over `format_error/1`. What about 

```elixir
form_error(Meta, File, ?MODULE, {erlang_error, Module, Desc})
```

?

Comment (josevalim) [inline lib/elixir/src/elixir_erl_compiler.erl]: The indirection here still feels really weird to me. I have a hard time understanding it. We still need compile_error, so let's continue using it here.

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**